aiohttp/web_urldispatcher.py

```diff
@@ -979,12 +986,39 @@ def __init__(self) -> None:
         super().__init__()
         self._resources: List[AbstractResource] = []
         self._named_resources: Dict[str, AbstractResource] = {}
+        self._resource_index: dict[str, list[AbstractResource]] = {}
+        self._matched_sub_app_resources: List[MatchedSubAppResource] = []
 
     async def resolve(self, request: Request) -> UrlMappingMatchInfo:
-        method = request.method
+        url_parts = request.rel_url.raw_parts
+        resource_index = self._resource_index
         allowed_methods: Set[str] = set()
 
-        for resource in self._resources:
+        # Walk the url parts looking for candidates. We walk the url backwards
+        # to ensure the most explicit match is found first. If there are multiple
+        # candidates for a given url part because there are multiple resources
+        # registered for the same canonical path, we resolve them in a linear
+        # fashion to ensure registration order is respected.
+        for i in range(len(url_parts), 0, -1):
+            url_part = "/" + "/".join(url_parts[1:i])
```

Comment:
And, maybe too complicated, but a tree structure might allow bailing out quickly. Maybe it would look something like:
```
candidates = [tree := tree.get(p) for p in url_parts if tree]
for tree in candidates[::-1]:
    for c in tree[None]:  # str keys are used for the prefix matching, None contains candidates?
```

Reply:
I'm just throwing around ideas here, feel free to ignore them. But, might be interesting to see if there's some further performance gains to be had.

Reply:
optimized in 5810f02750c88dfa6df7c3166eb5c0242f2c502c

Reply:
before
```
resolve_time_first_url: 0.002494540996849537
resolve_time_last_url: 0.0025858330773189664
resolve_time_miss: 0.007047250051982701
```

after
```
resolve_time_first_url: 0.0019859170133713633
resolve_time_last_url: 0.0020277079893276095
resolve_time_miss: 0.00666537499637343
```

Reply:
Looks like a little too much .. tests are failing now..

Reply:
Needed to use the `raw_path`

Reply:
and that seems to be a bit faster as well
```
resolve_time_first_url: 0.0018852080102078617
resolve_time_last_url: 0.0018728750001173466
resolve_time_miss: 0.005776290985522792
```

Reply:
Nice, so looks like it may be performing better than the original code even in the first_url condition now.

Reply:
Thanks for pushing me to optimize it more. Almost nobody ever does that in reviews.

I'm working on down streaming the changes today so I can validate in my production

Reply:
By the way, I only just realised while I was thinking about it overnight, but the benchmarks are biased.
Rather than first/last URL, it should really be best/worst time.
For the old code, the worst time will be the last URL. But, for the new code the worst time will be the one that is most nested and needs to go through multiple checks. The miss time is probably a reasonable estimate of the worst case, though it really dependent on the layout of a given app.

Reply:
I added 7 url case and 5002 url case.

The new code still beats the old 7 url case (although it doesn't make much difference)

I guess we could probably optimize the 404 handler a bit more

Reply:
I sped up the 404 case quite a bit

Reply:
before (5002 urls)
```
resolve_time_first_url: 0.0016546659753657877
resolve_time_last_url: 2.253484041953925
resolve_time_mid_size_url: 2.1529546250239946
resolve_time_long_url: 2.136738500033971
resolve_time_miss: 2.152159374963958
resolve_time_long_miss: 2.173428333015181
```
before (7 urls)
```
resolve_time_first_url: 0.0019221249967813492
resolve_time_last_url: 0.0038797500310465693
resolve_time_mid_size_url: 0.004664750013034791
resolve_time_long_url: 0.004066875029820949
resolve_time_miss: 0.00846966594690457
resolve_time_long_miss: 0.008330333046615124
```

after (5002 urls)
```
resolve_time_first_url: 0.001852833025623113
resolve_time_last_url: 0.0018547910149209201
resolve_time_mid_size_url: 0.0019565830007195473
resolve_time_long_url: 0.001857625029515475
resolve_time_miss: 0.0009622920188121498
resolve_time_long_miss: 0.005029833992011845
```
after (7 urls)
```
resolve_time_first_url: 0.001815625000745058
resolve_time_last_url: 0.0018432079814374447
resolve_time_mid_size_url: 0.001861582975834608
resolve_time_long_url: 0.0017958339885808527
resolve_time_miss: 0.0009489589720033109
resolve_time_long_miss: 0.005105625023134053
```

Reply:
Looks like I'm going to have to revert it since the web.Application handler always adds the app every time and than throws away the handler and refactoring that would probably be outside of scope for this PR